lib/__tests__/watch.test.ts

```diff
@@ -0,0 +1,191 @@
+
+    (uploadFolder as jest.Mock).mockResolvedValueOnce([]);
+
+    await watch(accountId, src, dest, options, postInitialUploadCallback);
+
+    expect(uploadFolder).toHaveBeenCalledWith(
+      accountId,
+      src,
+      dest,
+      {},
+      options.commandOptions,
+      options.filePaths,
+      options.mode
+    );
+    expect(postInitialUploadCallback).toHaveBeenCalled();
+  });
+
+  it('should log ready when watcher is ready', () => {
+    const accountId = 123;
+    const src = 'src-folder';
+    const dest = 'dest-folder';
+    const options = {
+      mode: MODE.draft,
+      remove: false,
+      disableInitial: true,
+      notify: '',
+      commandOptions: {},
+      filePaths: [],
+    };
+
+    watch(accountId, src, dest, options);
+
+    expect(chokidarMock.on).toHaveBeenCalledWith('ready', expect.any(Function));
+    chokidarMock.on.mock.calls[0][1]();
+    expect(logger.log).toHaveBeenCalledWith(expect.stringContaining('ready'));
+  });
+
+  it('should upload file when file is added', () => {
+    const accountId = 123;
+    const src = 'src-folder';
+    const dest = 'dest-folder';
+    const options = {
+      mode: MODE.draft,
+      remove: false,
+      disableInitial: true,
+      notify: '',
+      commandOptions: {},
+      filePaths: [],
+    };
+
+    watch(accountId, src, dest, options);
+
+    expect(chokidarMock.on).toHaveBeenCalledWith('add', expect.any(Function));
+    const addCallback = chokidarMock.on.mock.calls[1][1];
+    const filePath = '/some-file-path.html';
+
+    addCallback(filePath);
+
+    expect(logger.debug).toHaveBeenCalledWith(
+      expect.stringContaining('Attempting to upload')
+    );
+  });
+
+  it('should handle file change event and upload file', () => {
+    const accountId = 123;
+    const src = 'src-folder';
+    const dest = 'dest-folder';
+    const options = {
+      mode: MODE.draft,
+      remove: false,
+      disableInitial: true,
+      notify: '',
+      commandOptions: {},
+      filePaths: [],
+    };
+
+    watch(accountId, src, dest, options);
+
+    expect(chokidarMock.on).toHaveBeenCalledWith(
+      'change',
+      expect.any(Function)
+    );
+    const changeCallback = chokidarMock.on.mock.calls[2][1];
+    const filePath = 'changed-file-path.html';
+
+    changeCallback(filePath);
+
+    expect(logger.debug).toHaveBeenCalledWith(
+      expect.stringContaining('Attempting to upload')
+    );
+  });
+
+  it('should handle file delete event', () => {
+    const accountId = 123;
+    const src = 'src-folder';
+    const dest = 'dest-folder';
+    const options = {
+      mode: MODE.draft,
+      remove: true,
+      disableInitial: true,
+      notify: '',
+      commandOptions: {},
+      filePaths: [],
+    };
+
+    watch(accountId, src, dest, options);
+
+    expect(chokidarMock.on).toHaveBeenCalledWith(
+      'unlink',
+      expect.any(Function)
+    );
+    const deleteCallback = chokidarMock.on.mock.calls[2][1];
```

Comment:
I'm having a little trouble following exactly how this is testing the delete. Can you explain how this is working?

My understanding is that:
1. We're starting up the watch command
2. We're checking to make sure that an `unlink` listener was registered with chokidar
Those make sense to me

3. We're grabbing the mocked chokidar callback func, then manually executing it
4. We're making sure that the delete log shows up
This is the part I'm not following. If we're manually triggering the chokidar functions, then are we actually testing that chokidar is doing what we expect it to do? I'd imagine that it would be more effective for us to use fs or something to actually delete a file from within the watched folder to make sure the function subscribed to chokidar triggers.

Let me know if I'm just misinterpreting this though!

Reply:
https://github.com/HubSpot/hubspot-local-dev-lib/pull/195/commits/a54efdd0207005059c9f34db23cfaa33e8dc5de5